Kubernetes as a Composite config source

arjavdongaonkar opened the pull request:
Enables Kubernets as a type in composite config source

arjavdongaonkar commented on spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configserver/src/main/java/org/springframework/cloud/kubernetes/configserver/KubernetesEnvironmentRepositoryFactory.java:
```diff
@@ -0,0 +1,24 @@
+public class KubernetesEnvironmentRepositoryFactory
+	implements EnvironmentRepositoryFactory<KubernetesEnvironmentRepository, KubernetesConfigServerProperties> {
+
+	private final ObjectProvider<KubernetesEnvironmentRepository> kubernetesEnvironmentRepositoryProvider;
```
`ObjectProvider` allows lazy initialization.  It’s used to lazily retrieve `KubernetesEnvironmentRepository` when needed, avoiding early instantiation and unnecessary bean creation in the factory.
Since `KubernetesEnvironmentRepository` Bean has `@Profile("kubernetes")` annotation over it, so ObjectProvider is just to avoid NPE

arjavdongaonkar replied:
You're right that `build()` could return `null`. The factory itself doesn't have `@Profile("kubernetes")` because it might be used in other config sources (such as composite). However, would there ever be a case where the profile is not `kubernetes` when starting the Kubernetes Config Server?

ryanjbaxter replied:
You could technically start the Kubernetes Config Sever without the `kubernetes` profile.  One solution here would to make the `KubernetesEnvironmentRepositoryFactory` Bean conditional on the presence of a `KubernetesEnvironmentRepository`.

arjavdongaonkar replied:
Totally agree.

arjavdongaonkar commented:
> I had an initial pass over it, but I do not know all the details of what is going on. I might look again towards the end of the week... and btw, you have no tests here :)

Thanks! I'll add the tests. For context, please check this [issue](https://github.com/spring-cloud/spring-cloud-kubernetes/issues/1867).
tbh, This is my first open-source contribution, so I appreciate the feedback—I'll keep these points in mind.

ryanjbaxter commented:
It would be great to have some tests that actually test that the kubernetes environment repository works in a composite scenario.

arjavdongaonkar commented:
>It would be great to have some tests that actually test that the kubernetes environment repository works in a composite scenario.

Hey @ryanjbaxter,
Just to confirm, when we say "kubernetes environment repository works in a composite scenario," do we mean writing Composite Integration Tests where `kubernetes` is one of the type in the composite setup?

ryanjbaxter commented:
I am rethinking this a bit....

There are two ways to configure a composite environment repository

The first way you would do

```
spring:
  profiles:
    active: composite
  cloud:
    config:
      server:
        composite:
        - type: git
           uri: file:///path/to/rex/git/repo
        - type: kubernetes
          ```
          
 But I don't think this would work with this PR because there is not `kubernetes` profile active.
 
 The second way
 ```
 spring:
  profiles:
    active: git, kubernetes
  cloud:
    config:
      server:
        git:
          uri: file:///path/to/git/repo
          order: 2
          ```
 This should work fine I think

arjavdongaonkar commented:
>This should work fine I think

Sure, Will check this out
   - As far as I know, the `kubernetes` profile is automatically included when the config server runs in a Kubernetes cluster, so there’s no need to explicitly define it in the `spring.profiles.include / active` property.  This beahviour is due to [this](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/f1c7e9d8babe41f297d6162a352e8451f4b6ff6e/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/profile/AbstractKubernetesProfileEnvironmentPostProcessor.java#L79) code
   - The main advantage of using `composite` is the ability to define multiple instances of a profile.  
   - For example, my use case requires multiple Vault instances (application and infra) with different keys, along with Kubernetes and Git sources:  

   ```yaml
   spring:
     profiles:
       active: composite
     cloud:
       config:
         server:
           composite:
             - type: kubernetes
             - type: vault
               defaultKey: application
             - type: vault
               defaultKey: infra
             - type: git
       kubernetes:
         secrets:
           enableApi: true
   ```  
   - Additionally, This configuration allows me to control the order of property sources.
   - The above setup is working as expected in my environment and use case with code of this PR.  

Do we have a specific reason to prefer the second approach over the composite model?

ryanjbaxter commented:
> Sure, Will check this out

Please add a test for this

> As far as I know, the kubernetes profile is automatically included when the config server runs in a Kubernetes cluster, so there’s no need to explicitly define it in the spring.profiles.include / active property. This beahviour is due to [this](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/f1c7e9d8babe41f297d6162a352e8451f4b6ff6e/spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/profile/AbstractKubernetesProfileEnvironmentPostProcessor.java#L79) code

I missed this in your tests but now I see it, sorry.

> Do we have a specific reason to prefer the second approach over the composite model?

No but both are valid and wanted to make sure it is tested.

arjavdongaonkar commented:
Hey @ryanjbaxter 

While writing a test case for *Scenario 2* of the composite configuration (comma-separated active profiles), I found that the *composite* environment repository [sorts](https://github.com/spring-cloud/spring-cloud-config/blob/4e128da59437ce68fd0f2bd27c50c366ebd4a9e3/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/CompositeEnvironmentRepository.java#L54) the repositories based on the `order` property. However, in the case of the *Kubernetes* environment repository, the ordering was not happening because we never defined an `order` property for it.  

Taking reference of similar profile types in *Spring Cloud Config Server* (e.g., [Vault](https://github.com/spring-cloud/spring-cloud-config/blob/4e128da59437ce68fd0f2bd27c50c366ebd4a9e3/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AbstractVaultEnvironmentRepository.java#L78), [Native](https://github.com/spring-cloud/spring-cloud-config/blob/4e128da59437ce68fd0f2bd27c50c366ebd4a9e3/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/NativeEnvironmentRepository.java#L98), [Git](https://github.com/spring-cloud/spring-cloud-config/blob/4e128da59437ce68fd0f2bd27c50c366ebd4a9e3/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AbstractScmEnvironmentRepository.java#L52) ), I found that there, we pass the `order` property via `properties` when creating their environment repository beans. However, the Kubernetes environment repository bean was missing this argument.  

I initially considered accepting `order` as an environment variable and injecting it using `@Value`, but this felt like a workaround. 
For a consistent approach and proper ordering, I have:  
1. Introduced an `KubernetesConfigServerProperties` argument in the *Kubernetes environment repository bean* (following the pattern used by other profiles).  
2. Added relevant test cases to verify the expected ordering behaviour.  

Let me know if you have any concerns or if you'd like to discuss alternative approaches!

ryanjbaxter commented on spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configserver/src/main/java/org/springframework/cloud/kubernetes/configserver/KubernetesEnvironmentRepository.java:
```diff
@@ -46,11 +47,15 @@ public class KubernetesEnvironmentRepository implements EnvironmentRepository {
 
+	private int order;
+
 	public KubernetesEnvironmentRepository(CoreV1Api coreApi,
```
We need to keep the old constructor around, but mark it as `@Deprecated` and set `this.order` to `KubernetesConfigServerProperties.DEFAULT`.

arjavdongaonkar replied:
Done
